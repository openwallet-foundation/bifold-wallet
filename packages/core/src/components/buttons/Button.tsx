import React, { forwardRef, useState } from 'react'
import { TouchableOpacity, View } from 'react-native'

import { useTheme } from '../../contexts/theme'

import { Button, ButtonType, ButtonProps } from './Button-api'
import { ThemedText } from '../texts/ThemedText'

const ButtonImplComponent = (
  {
    title,
    buttonType,
    accessibilityLabel,
    accessibilityHint,
    testID,
    onPress,
    disabled = false,
    maxfontSizeMultiplier,
    children,
  }: ButtonProps,
  ref: React.LegacyRef<TouchableOpacity>
) => {
  const { Buttons, heavyOpacity } = useTheme()
  const buttonStyles = {
    [ButtonType.Critical]: {
      color: Buttons.critical,
      colorDisabled: Buttons.criticalDisabled,
      text: Buttons.criticalText,
      textDisabled: Buttons.criticalTextDisabled,
    },
    [ButtonType.Primary]: {
      color: Buttons.primary,
      colorDisabled: Buttons.primaryDisabled,
      text: Buttons.primaryText,
      textDisabled: Buttons.primaryTextDisabled,
    },
    [ButtonType.Secondary]: {
      color: Buttons.secondary,
      colorDisabled: Buttons.secondaryDisabled,
      text: Buttons.secondaryText,
      textDisabled: Buttons.secondaryTextDisabled,
    },
    [ButtonType.Tertiary]: {
      color: Buttons.tertiary,
      colorDisabled: Buttons.tertiaryDisabled,
      text: Buttons.tertiaryText,
      textDisabled: Buttons.tertiaryTextDisabled,
    },
    [ButtonType.ModalCritical]: {
      color: Buttons.modalCritical,
      colorDisabled: Buttons.modalCriticalDisabled,
      text: Buttons.modalCriticalText,
      textDisabled: Buttons.modalCriticalTextDisabled,
    },
    [ButtonType.ModalPrimary]: {
      color: Buttons.modalPrimary,
      colorDisabled: Buttons.modalPrimaryDisabled,
      text: Buttons.modalPrimaryText,
      textDisabled: Buttons.modalPrimaryTextDisabled,
    },
    [ButtonType.ModalSecondary]: {
      color: Buttons.modalSecondary,
      colorDisabled: Buttons.modalSecondaryDisabled,
      text: Buttons.modalSecondaryText,
      textDisabled: Buttons.modalSecondaryTextDisabled,
    },
    [ButtonType.ModalTertiary]: {
      color: Buttons.modalTertiary,
      colorDisabled: Buttons.modalTertiaryDisabled,
      text: Buttons.modalTertiaryText,
      textDisabled: Buttons.modalTertiaryTextDisabled,
    },
  }
  const [isActive, setIsActive] = useState<boolean>(false)

  return (
    <TouchableOpacity
      onPress={onPress}
      accessible
      accessibilityLabel={accessibilityLabel}
      accessibilityHint={accessibilityHint}
      accessibilityRole={'button'}
      onPressIn={() => setIsActive(!disabled && true)}
      onPressOut={() => setIsActive(false)}
      testID={testID}
      style={[
        buttonStyles[buttonType].color,
        disabled && buttonStyles[buttonType].colorDisabled,
        isActive &&
          (buttonType === ButtonType.Secondary || buttonType === ButtonType.Tertiary) && {
            backgroundColor: Buttons.primary.backgroundColor,
          },
      ]}
      disabled={disabled}
      activeOpacity={heavyOpacity}
      ref={ref}
    >
      <View
        style={{
          flexDirection: 'row',
          justifyContent: 'center',
          alignItems: 'center',
        }}
      >
        {children}
        <ThemedText
          maxFontSizeMultiplier={maxfontSizeMultiplier}
          style={[
            buttonStyles[buttonType].text,
            disabled && buttonStyles[buttonType].textDisabled,
            isActive && { textDecorationLine: 'underline' },
            isActive && buttonType === ButtonType.Secondary && { color: Buttons.primaryText.color },
          ]}
        >
          {title}
        </ThemedText>
      </View>
    </TouchableOpacity>
  )
}

const ButtonImpl = forwardRef<TouchableOpacity, ButtonProps>(ButtonImplComponent)
export default ButtonImpl
export { ButtonType, ButtonImpl }
export type { Button, ButtonProps }
