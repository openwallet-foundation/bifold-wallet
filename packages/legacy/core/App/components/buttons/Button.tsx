import React, { forwardRef, useState } from 'react'
import { TouchableOpacity, View } from 'react-native'

import { useTheme } from '../../contexts/theme'

import { Button, ButtonType, ButtonProps } from './Button-api'
import { ThemedText } from '../texts/ThemedText'

const ButtonImplComponent = (
  {
    title,
    buttonType,
    accessibilityLabel,
    accessibilityHint,
    testID,
    onPress,
    disabled = false,
    children,
  }: ButtonProps,
  ref: React.LegacyRef<TouchableOpacity>
) => {
  const { Buttons, heavyOpacity } = useTheme()
  const buttonStyles = {
    [ButtonType.Critical]: { color: Buttons.critical, text: Buttons.primaryText },
    [ButtonType.Primary]: { color: Buttons.primary, text: Buttons.primaryText },
    [ButtonType.Secondary]: { color: Buttons.secondary, text: Buttons.secondaryText },
    [ButtonType.Tertiary]: { color: Buttons.tertiary, text: Buttons.tertiaryText },
    [ButtonType.ModalCritical]: { color: Buttons.modalCritical, text: Buttons.primaryText },
    [ButtonType.ModalPrimary]: { color: Buttons.modalPrimary, text: Buttons.modalPrimaryText },
    [ButtonType.ModalSecondary]: { color: Buttons.modalSecondary, text: Buttons.modalSecondaryText },
    [ButtonType.ModalTertiary]: { color: Buttons.modalTertiary, text: Buttons.modalTertiaryText },
  }
  const [isActive, setIsActive] = useState<boolean>(false)

  return (
    <TouchableOpacity
      onPress={onPress}
      accessible
      accessibilityLabel={accessibilityLabel}
      accessibilityHint={accessibilityHint}
      accessibilityRole={'button'}
      onPressIn={() => setIsActive(!disabled && true)}
      onPressOut={() => setIsActive(false)}
      testID={testID}
      style={[
        buttonStyles[buttonType].color,
        disabled &&
          (buttonType === ButtonType.Primary
            ? Buttons.primaryDisabled
            : buttonType === ButtonType.Secondary
            ? Buttons.secondaryDisabled
            : Buttons.tertiaryDisabled),
        isActive &&
          (buttonType === ButtonType.Secondary || buttonType === ButtonType.Tertiary) && {
            backgroundColor: Buttons.primary.backgroundColor,
          },
      ]}
      disabled={disabled}
      activeOpacity={heavyOpacity}
      ref={ref}
    >
      <View
        style={{
          flexDirection: 'row',
          justifyContent: 'center',
          alignItems: 'center',
        }}
      >
        {children}
        <ThemedText
          style={[
            buttonStyles[buttonType].text,
            disabled &&
              (buttonType === ButtonType.Primary
                ? Buttons.primaryTextDisabled
                : buttonType === ButtonType.Secondary
                ? Buttons.secondaryTextDisabled
                : Buttons.tertiaryTextDisabled),
            isActive && { textDecorationLine: 'underline' },
            isActive && buttonType === ButtonType.Secondary && { color: Buttons.primaryText.color },
          ]}
        >
          {title}
        </ThemedText>
      </View>
    </TouchableOpacity>
  )
}

const ButtonImpl = forwardRef<TouchableOpacity, ButtonProps>(ButtonImplComponent)
export default ButtonImpl
export { ButtonType, ButtonImpl }
export type { Button, ButtonProps }
