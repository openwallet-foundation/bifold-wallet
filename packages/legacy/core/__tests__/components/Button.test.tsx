import { render } from '@testing-library/react-native'
import React from 'react'

// eslint-disable-next-line import/no-named-as-default
import Button, { ButtonType } from '../../App/components/buttons/Button'

describe('Button Component', () => {
  test('Primary renders correctly', () => {
    const tree = render(
      <Button
        title={'Hello Primary'}
        accessibilityLabel={'primary'}
        onPress={() => {
          return
        }}
        buttonType={ButtonType.Primary}
      />
    )

    expect(tree).toMatchSnapshot()
  })

  test('Secondary renders correctly', () => {
    const tree = render(
      <Button
        title={'Hello Secondary'}
        accessibilityLabel={'secondary'}
        onPress={() => {
          return
        }}
        buttonType={ButtonType.Secondary}
      />
    )

    expect(tree).toMatchSnapshot()
  })
})
