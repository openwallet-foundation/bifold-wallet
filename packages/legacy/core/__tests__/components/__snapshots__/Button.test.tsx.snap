// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component Primary renders correctly 1`] = `
<View
  accessibilityLabel="primary"
  accessibilityRole="button"
  accessibilityState={
    Object {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    Object {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "backgroundColor": "#42803E",
      "borderRadius": 4,
      "opacity": 1,
      "padding": 16,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      maxFontSizeMultiplier={2}
      style={
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 18,
            "fontWeight": "normal",
          },
          Array [
            Object {
              "color": "#FFFFFF",
              "fontSize": 18,
              "fontWeight": "bold",
              "textAlign": "center",
            },
            false,
            false,
            false,
          ],
        ]
      }
    >
      Hello Primary
    </Text>
  </View>
</View>
`;

exports[`Button Component Secondary renders correctly 1`] = `
<View
  accessibilityLabel="secondary"
  accessibilityRole="button"
  accessibilityState={
    Object {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    Object {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "borderColor": "#42803E",
      "borderRadius": 4,
      "borderWidth": 2,
      "opacity": 1,
      "padding": 16,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      maxFontSizeMultiplier={2}
      style={
        Array [
          Object {
            "color": "#FFFFFF",
            "fontSize": 18,
            "fontWeight": "normal",
          },
          Array [
            Object {
              "color": "#42803E",
              "fontSize": 18,
              "fontWeight": "bold",
              "textAlign": "center",
            },
            false,
            false,
            false,
          ],
        ]
      }
    >
      Hello Secondary
    </Text>
  </View>
</View>
`;
